### ISSUE TITLE
Timing issue when using a Factory with async taskexecutor

### ISSUE THREAD

**Issue by cvwj** [spring-projects/spring-statemachine#298]
I use the SSM 1.2.0.RELEASE together with Spring Boot 2.0.0.BUILD-SNAPSHOT.

In my setup, I start a new State Machine instance every time I receive a request. I'm using a StateMachineFactory to get a machine. The Factory is configured like this:
```
    @Override
    public void configure(StateMachineConfigurationConfigurer<String, String> config) throws Exception {
        config
                .withConfiguration()
                .autoStartup(true)
                .taskExecutor(new SimpleAsyncTaskExecutor());
    }

```

When the request arrives in my REST service, I do this:

```
        StateMachine<String, String> machine = leadAMachineFactory.getStateMachine();
        Message<String> leadMessage = MessageBuilder
                .withPayload("lead_arrived")
                .setHeader("lead", leadA)
                .build();

        machine.sendEvent(leadMessage);
```

This results in a NullPointerException in AbstractStateMachine.acceptEvent(). The reason is that the machine.currentState is null when I send the Event to it. It is not finished initialising.

If I instead wait with sending the event until machine is in a valid state, everything works fine:

```
        StateMachine<String, String> machine = leadAMachineFactory.getStateMachine();
        Message<String> leadMessage = MessageBuilder
                .withPayload("lead_arrived")
                .setHeader("lead", leadA)
                .build();

        new Thread(new Runnable() {
            @Override
            public void run() {
                while (machine.getState() == null) {
                    try {
                        System.out.println("Machine not ready...");
                        Thread.sleep(100);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
                machine.sendEvent(leadMessage);
            }
        }).start();
```

If the Factory is configured with a Synchronous task Executor like the default one, there are also no problems, as the sendEvent statement is executed in the same thread, but at a later point than the initialising code.

One (naive?) solution could be to block the sendEvent() until the machine is in a ready condition.

**Comment by jvalkeal** [comment]
Thanks for pointing this out. I'll poke around to find some kind of fix for this. In a meanwhile I suppose starting machine manually and using `StateMachineListener` to know when started event is send would do a trick.

I'm also thinking that if autostart is used then maybe internal factory should should not return a machine before it is started. That said, user should not have to worry about these things.

**Comment by jvalkeal** [comment]
When we find a good fix for 1.2.1 it's definitely something to backport to 1.1.x.

**Comment by jvalkeal** [comment]
Fixed per https://github.com/spring-projects/spring-statemachine/commit/8bb852cc7907e8d53dfde2a860b86e0722d2eccf in branch 1.2.x.
